WSUCptSCapstone-Fall2022Spring2023/psd-gamifiedapp
# Sprint1/core iron python integration

closes #4 
closes #18 

Demo Video Link: [https://youtu.be/-WFSqS2P4gg](https://youtu.be/-WFSqS2P4gg)

**Overview:**

This pull request implements the core functionality of IronPython into the Unity project. This means setting up the project so that user code can be safely run in a managed environment while interacting with lesson scripts generated by us. It also addresses the emergency fix of the gitignore. While this would normally be handled by a separate pull request. It is vitally important these changes are merged ASAP, as without them all pull requests are lagging and git operations are slowed down.

**Details:**

This pull request features the addition of 4 very important files, and updates to 1 file.

LevelDefinition.cs: This file defines the data contained in a level in relation to the IronPython implementation. This file will be expanded on later to interface with other components. This PR adds the initial version of this file with the information needed by the IronPython implementation.

IronPythonContainer.cs: This file is the container that holds the IronPython implementation and primarily handles the interface between the IronPython interpreter and the main Unity application. This PR adds that file and fully implements it.

CRTShader.shader: This HLSL file contains the shader code for adding a CRT shader to the main camera. This was added in this PR mostly to improve the quality of the demo and to set a foundation for some future work. It was included in this PR even though it was developed in this branch because of the emergency requirement to fold in the branch that fixes the .gitignore.

CRTPostProcess.cs: A C# script that interfaces the shader script with the Unity interpreter and applies it to the main camera.

.gitignore: CRITICALLY IMPORTANT UPDATES. This PR includes the emergency patch for the gitignore. That should hopefully make this the only PR that has this horribly high number of file changes. Unfortunately there is no way to fix that without this PR.

**Testing:**

We have yet to formalize testing plans for the Unity project or to set up Unity's internal unit test framework. As such we are still currently relying on manual testing . My current manual test process is as follows.

1. Start Project
2. Simulate without writing any code, check that it exits sim with code 1 when testVar hits 20
3. Simulate with code testVar-=6, check that testVar behaves as expected and exits with code 0 when testVar hits -20.
4. Try using disallowed variables, check that exception is thrown
5. Try using import, check exception is thrown.
